first commit project 1 group 2

xaviersellam opened the pull request:
Hi,
Here is the beginning of our project but we have few problems:
-we tried to create the classic BC model (the easiest) but it doesn't work: indeed, the arguments S, K, r... are taken randomly even if we fix them before running the code (we fix them in constantblackscholesprocess.cpp)
Furthermore, we do not understant what is the  StochasticProcess1D (constantBSProcess is supposed to inherit from it)

Thank you in advance for you help and sorry for the late return, we are not very familiar with c++

lballabio commented on project1/constantblackscholesprocess.cpp:
```diff
@@ -1,3 +1,55 @@
 #include "constantblackscholesprocess.hpp"
 
+ConstantBSProcess::ConstantBSProcess(){
+double S = 100.0;
```
If you write `double S = 100.0` you're not assigning a value to the member of the instance; you're declaring a new variable with the same name which will be destroyed at the end of the `}`.  This is why you get random values in your variables: the memory was not initialized.

Instead, write `S = 100.0` without the `double`.

lballabio commented on project1/constantblackscholesprocess.cpp:
```diff
@@ -1,3 +1,55 @@
+
+double ConstantBSProcess :: drift(){
+std::cout<<r<<std::endl;
+return((r -0.5*v*v)*T);}
```
You don't need to multiply by T; this method only returns the drift coefficient.

lballabio commented on project1/constantblackscholesprocess.cpp:
```diff
@@ -1,3 +1,55 @@
+
+double ConstantBSProcess :: diffusion(){
+    double k=gaussian_box_muller();
+return(v*pow(T,0.5)*k);}
```
Same here, no need for `k` and `T`.

lballabio commented on project1/constantblackscholesprocess.hpp:
```diff
@@ -1,12 +1,32 @@
+double norm_cdf(const double& x);
+double norm_pdf(const double& x);
+
+    class ConstantBSProcess {//: public StochasticProcess1D {
```
`StochasticProcess1D` is another class declared in `ql/stochasticprocess.hpp` (see <https://github.com/lballabio/QuantLib/blob/master/ql/stochasticprocess.hpp#L162>).  In order to use your class in the Monte Carlo model implemented in QuantLib, you need to inherit from it and implement its pure virtual methods.  If you're not familiar with C++, they are those declared with a `= 0` at the end; you'll learn about them in your C++ course when you take it.  In this case, `virtual Real x0() const = 0;` (line 178), `virtual Real drift(Time t, Real x) const = 0;` (line 180), and `virtual Real diffusion(Time t, Real x) const = 0;` (line 184).  In order to implement them in your class, you need to declare them in the exact same way but without the `=0` and to give them a body. See below.

lballabio commented on project1/constantblackscholesprocess.hpp:
```diff
@@ -1,12 +1,32 @@
+
+    public:
+ConstantBSProcess();
+double diffusion();
```
This should be `Real diffusion(Time t, Real x) const`.  You're not going to use `t` or `x` in the body (because your process has constant parameters, so they don't depend on x or t) but you need to replicate the signature of the method declared in the base class.

lballabio commented on project1/constantblackscholesprocess.hpp:
```diff
@@ -1,12 +1,32 @@
+    public:
+ConstantBSProcess();
+double diffusion();
+double drift();
```
Same here: this need to be `Real drift(Time t, Real x) const`.

lballabio commented:
How is it going? Do you need help?

xaviersellam commented:
Hi, with the coronavirus problem and the move that we had to do rapidly, we didn’t went much further, we will do it during next week.

Thank you, we will tell you when we face problems


Le 21 mars 2020 à 10:43, Luigi Ballabio <notifications@github.com> a écrit :

﻿

How is it going? Do you need help?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/lballabio/IMT2020/pull/9#issuecomment-602020276>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOW7TITITT37YXGIXSDTNKDRISD4HANCNFSM4LGNHL4Q>.

xaviersellam commented:
Hello,

Firstly, we hope that you and your relatives are in good health.



Here's a summary of what we have done (with some questions).

1) We have taken your remarks into account and corrected our mistakes. 

2) We have added a boolean to our MCVanillaEngine: if it returns true we will use a ConstantBSProcess for the pricing, otherwise we will use the existing path generator/path pricer. We have simply created a boolean and implemented a condition if/else in the path generator method.

3) We want to extract from the handles in the original process the parameters risk-free rate, dividend yield and volatility provided as paramaters of the class GeneralizedBlackScholesProcess.
Thanks to Getters we can manipulate the handles but how can we extract our constant parameters from them ? Are these handles vectors giving time-depedent parameters ? Have we to calculate their means and use them as                       
constant parameters for our ConstantBSProcess ?  
Other question, do we use the local volatility or the global one ?

4)What's the difference between process and process_ in MCEuropeanEngine ?

5) We have difficulties in importing the Quantlib library. Indeed our "#include.." are often not recognized by our algorithm. How can we fix that ?



We have spent a lot time working on the code to understand it, we hope we're making some sense and our questions are relevant.

PS: the first second commit is irrelevant so check only the second "second commit".

Thank you,
Nazim and Xavier

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,29 @@
+ConstantBSProcess(double S, double K, double r, double v, double T, double q);
+path_generator_2(const double& S, const double& K, const double& r, const double& v, const double& T) ;
+virtual Real diffusion(Time t, Real x);
+virtual Real drift(Time t,Real x);
```
These two methods need to be `const` like the ones in `StochasticProcess1D` or the compiler won't match them.
You'll also need the `x0` method.

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,29 @@
+
+    public:
+ConstantBSProcess(double S, double K, double r, double v, double T, double q);
+path_generator_2(const double& S, const double& K, const double& r, const double& v, const double& T) ;
```
You don't need a path generator.  See the comment in the engine code.

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,29 @@
+class ConstantBSProcess{ : public StochasticProcess1D {
+    private:
+double S;
+double K;
```
You don't need K or T.

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,29 @@
+
+namespace QuantLib {
+
+class ConstantBSProcess{ : public StochasticProcess1D {
```
Extra `{`:
```suggestion
class ConstantBSProcess : public StochasticProcess1D {
```

xaviersellam commented:
Hello,


We have taken in account your remarks and we have new questions:

1) We hesitate for the type of the dividend q : is it a rate like r or a spread ?

2) We have tried many times to import the Quantlib library without results in Codeblocks. Therefore we can't run our programm ...

3) We have put all the parameters and created our generalizedblackscholesprocess and european engine in order to use
its pricing method to evaluate our derivative. Could you check if our code is good ?

4) Finally, in order to calculate the accuracy we have to calculate the result of our algorithm with the real value of the option:                 
is this real value equal to the output of the NPV method of the class McVanilla option ?


Thanks,

lballabio commented:
The code is better.  There are still a few corrections, but I'd try to get your program running first.
What errors do you get when you try to compile in CodeBlocks?

lballabio commented:
Also, are you using Linux or Windows? Do you have any alternatives to CodeBlocks?

lballabio commented:
Also:
1. q is a Rate
4. the "real" value is the analytic one. You can get it using `AnalyticEuropeanEngine` with the option.

xaviersellam commented:
Thank you for your answer.
The first error is:
 ||=== Build: Debug in finance (compiler: GNU GCC Compiler) ===|
C:\Users\xavie\Downloads\QuantLib-1.18\QuantLib-1.18\ql\qldefines.hpp|38|fatal error: boost/config.hpp: No such file or directory|
The second error is:
C:\Users\xavie\OneDrive\Documents\Cours IMTA\S4\Finance empirique\C++\IMT2020\project1\main.cpp|2|fatal error: constantblackscholesprocess.hpp: No such file or directory|

The third one
C:\Users\xavie\OneDrive\Documents\Cours IMTA\S4\Finance empirique\C++\IMT2020\project1\stochasticprocess.cpp|22|fatal error: stochasticprocess.hpp: No such file or directory|

We downloaded quantlib to our computer and tried to put it in the compile path but it didn't work. We work under Windows.

There is maybe another way to run this code?

lballabio commented:
It seems you already set up QuantLib so that the compiler can find its header files, because the first error happens while compiling C:\Users\xavie\Downloads\QuantLib-1.18\QuantLib-1.18\ql\qldefines.hpp.  However, you probably need to compile the library, too.

This said:

- first error: you also need Boost (https://www.boost.org/).  You don't need to compile it: download it and decompress it.
- second error: you're #including "constantblackscholesprocess.hpp" but your file is called "ConstantBSProcess.h".
- third error: I think you also have to include your current directory in the include path.

However, if you're not comfortable with C++, fixing the errors can become a long process (once you fix the ones above, there will be others).  Is there any chance you can ask a course mate for some help with the language?

xaviersellam commented:
But I don't understand your third tips : 
"third error: I think you also have to include your current directory in the include path."

xaviersellam commented:
and we will try to fix the errorsalone because it is hard to explain someone else you code when you are not in front of him.

xaviersellam commented:
We fixed the previous errors. There are now the implementation errors that appear so we will look at this.

xaviersellam commented:
Hello,


We have managed to run our code by importing the quantlib library. We have corrected many errors but we have difficulties to <handle> 4 type of mistakes in our code :

1) the compiler does not recognize our class ConstantBSprocess : "'ConstantBSProcess' does not name a type" despite the fact we have well included its header in our files.

2) We have declared the boolean constparameters as an attribute of MCvanilla engine and McEuropean engine ( in their constructors) yet this boolean and its copy constparameters are not recognized when we try to create  dynamically this kind of engine.


3)When we create a LocalConstantVol we put the right kinds of parameters in the constructor but the compiler displays the error " not matching function for call to the constructor".


4)We do not understand this error from the original code : "#error This file requires compiler and library support for the ISO C++ 2011 standard.
This support is currently experimental, and must be enabled with the -std=c++11 or -std=gnu++11 compiler options..."


5) in the main, to calculate the elapsed time we put :

#include<chrono>
#include<ctime>

int main(){

std::chrono::time_point<std::chrono::system_clock> start , end ;
start=std::chrono::system_clock::now();
(our code)
end=std::chrono::system_clock::now();
std::cout << start-end << std::endl;

Yet the compiler displays that std::chrono,start and end are not defined. 
Could you help us with these errors?


Thanks,

lballabio commented:
I'll have a look.  Given the circumstances, I'm writing to Jean-Marc Le Caillec and asking him to contact you and give you an extension of the deadline.

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,66 @@
+#include <blackscholesprocess.hpp>
+#include <ConstantBSProcess.h>
+
+ConstantBSProcess::ConstantBSProcess(Real S,Rate r,Volatility v,Rate q){
+S = 100.0;
+r = process_->riskFreeRate()->zeroRate(grid.back(), Continuous);
+v = process_->blackVolatility()->blackVol(grid.back(), Continuous);
+q = process_->dividendYield()->zeroRate(grid.back(), Continuous);
+}
```
The idiomatic way to assign the passed parameters to the data members in C++ is:
```suggestion
ConstantBSProcess::ConstantBSProcess(Real S,Rate r,Volatility v,Rate q)
: S(S), r(r), v(v), q(q) {}
```
You don't have a process here; you'll extract the rates and volatility in the engine before passing them.

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,66 @@
+#include <blackscholesprocess.hpp>
+#include <ConstantBSProcess.h>
+
```
You defined `ConstantBSProcess` in namespace QuantLib (see the hpp) so the implementation also needs to be inside the same namespace.
```suggestion
namespace QuantLib {
```
You'll also need to close the namespace at the end of the file.

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,66 @@
+
+
+virtual ConstantBSProcess :: Real drift(Time t,Real x) const{
+return((r -0.5*v*v));}
```
This is the expression of the drift for the log of the underlying, which would require more logic in the process.  Instead, we can use the drift for the original geometric process:
```suggestion
return (r-q)*x;
}
```

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,66 @@
+
+virtual ConstantBSProcess :: Real diffusion(Time t, Real x) const{
+//    double k=gaussian_box_muller();
+return(v);}
```
Same here: in the geometric process, the vol is multiplied by the underlying.
```suggestion
return v*x;
}
```

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,27 @@
+Rate q;
+
+    public:
+ConstantBSProcess();
```
You don't need a default constructor, just the one that takes the arguments.

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,66 @@
+} **/
+
+
+virtual ConstantBSProcess :: Real drift(Time t,Real x) const{
```
The correct syntax is
```suggestion
Real ConstantBSProcess::drift(Time t,Real x) const{
```
Same for the others.

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,66 @@
+#include <blackscholesprocess.hpp>
```
You don't need this header here.

lballabio commented on project1/main.cpp:
```diff
@@ -1,19 +1,107 @@
+            boost::shared_ptr<BlackVolTermStructure>(
+                new BlackConstantVol(settlementDate, calendar, v,
+                                     dayCounter)));
+        Handle<LocalVolTermStructure> localVolTS(
```
You don't need to pass a local constant vol to the process.  You can delete this.

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,27 @@
+#include <cmath>
+#include <algorithm>
+#include <stochasticprocess.hpp>
```
This should be
```suggestion
#include <ql/stochasticprocess.hpp>
```
so that it finds the one inside your copy of QuantLib.  If it works, you can remove the `stochasticprocess.*` files in this directory.

lballabio commented on project1/main.cpp:
```diff
@@ -1,19 +1,107 @@
 #include <ql/pricingengines/vanilla/mceuropeanengine.hpp>
 #include <ql/quantlib.hpp>
+#include <mcvanillaengine.hpp>
+#include <blackscholesprocess.hpp>
+#include <stochasticprocess.hpp>
```
```suggestion
#include <ql/processes/blackscholesprocess.hpp>
#include <ql/stochasticprocess.hpp>
```

lballabio commented on project1/mcvanillaengine.hpp:
```diff
@@ -0,0 +1,185 @@
+            }
+            return ext::shared_ptr<path_generator_type>(
+                   new path_generator_type(process_, grid,
+                                           generator, brownianBridge_));}
```
One `}` too many.
```suggestion
                                           generator, brownianBridge_));
```

lballabio commented on project1/mcvanillaengine.hpp:
```diff
@@ -0,0 +1,185 @@
+        Size requiredSamples_, maxSamples_;
+        Real requiredTolerance_;
+        bool brownianBridge_;
+        BigNatural seed_;
```
You need to add another private data member to store `constantparameters`.

lballabio commented on project1/mcvanillaengine.hpp:
```diff
@@ -0,0 +1,185 @@
+                RNG::make_sequence_generator(dimensions*(grid.size()-1),seed_);
+
+            ext::shared_ptr<StochasticProcess1D> usedProcess;
+            if (constantparameters){
```
`process_` is a generic `StochasticProcess1D`, which doesn't have the methods `riskFreeRate` etc.
Before calling them, you need to downcast it by adding something like

    ext::shared_ptr<GeneralizedBlackScholesProcess> bsprocess = ext::dynamic_pointer_cast<GeneralizedBlackScholesProcess>(process_);

so below you can call `bsprocess->riskFreeRate()` etc.

lballabio commented on project1/mceuropeanengine.hpp:
```diff
@@ -116,7 +118,8 @@ namespace QuantLib {
              Size maxSamples,
-             BigNatural seed)
+             BigNatural seed,
+             bool constantparameters);
```
Extra ';'
```suggestion
             bool constantparameters)
```

lballabio commented on project1/mceuropeanengine.hpp:
```diff
@@ -80,6 +81,7 @@ namespace QuantLib {
         MakeMCEuropeanEngine_2& withMaxSamples(Size samples);
         MakeMCEuropeanEngine_2& withSeed(BigNatural seed);
         MakeMCEuropeanEngine_2& withAntitheticVariate(bool b = true);
+        MakeMCEuropeanEngine_2& withconstantparameters(bool constantparameters = false);
```
You need to give an implementation of this method (see how the others are done).  You also need a data member to store `constantparameters`.

xaviersellam commented:
Hello,

We have worked on the project this week end and  fixed all the errors.

We have one last new mistake that prevents us from running our code:
'ld returned l exit status'
'cannot find C:\\Users\xavie\Downloads\QuantLib-1.10\Quantlib-1.18: Permission denied.'


This mistakes appeared suddenly when we have corrected the other errors.
We looked for solutions and tried to fix it by closing codeblocks in the task manager 
in vain.

Thanks,

lballabio commented on project1/finance.cbp:
```diff
@@ -0,0 +1,51 @@
+		</Linker>
+		<Unit filename="ConstantBSProcess.cpp" />
+		<Unit filename="ConstantBSProcess.h" />
+		<Unit filename="blackscholesprocess.cpp" />
```
You might need to remove this one to avoid conflicts with the one in the library.

lballabio commented on project1/finance.cbp:
```diff
@@ -0,0 +1,51 @@
+			<Add option="-Wall" />
+		</Compiler>
+		<Linker>
+			<Add library="C:/Users/xavie/Downloads/QuantLib-1.18/QuantLib-1.18" />
```
You're adding a directory, not a library, so the linker complains (with that strange "permission error" message, which is not helpful).

Did you compile QuantLib when you downloaded it?  If you did, add the library here instead (it might be a .lib file, or a .so depending on how you compiled it).  If you didn't, do it now.  There's no project for CodeBlocks, but I think you can create a new project and tell CodeBlocks "add every cpp file in C:/Users/xavie/Downloads/QuantLib-1.18/QuantLib-1.18/ql".  You also need to tell CodeBlocks to create a library, not an executable.  As for this project, add C:/Users/xavie/Downloads/QuantLib-1.18/QuantLib-1.18 to the include directories.  Let me know if that works.

xaviersellam replied:
ok but how can we compile quantlib? 
I downloaded Quantib, I think I'm using it because when I include files it works but I never compiled anything and what I downloaded is too big to find something

lballabio replied:
Create a new project in CodeBlocks and add every `.cpp` file in `C:/Users/xavie/Downloads/QuantLib-1.18/QuantLib-1.18/ql`.  There's probably a way to add the whole directory at once, not file per file (they're too many).

You also need to tell CodeBlocks to create a library, not an executable. Like you did for your project, add `C:/Users/xavie/Downloads/QuantLib-1.18/QuantLib-1.18` to the include directories.

xaviersellam replied:
I did not find any way to add the whole directory, codeblocks only adds files that are in the same folder. 
I asked to my classmates if someone used Codeblocks but no answers at the moment... I will tell you if we achieve to go further.

Thank you for your time and your help.

lballabio replied:
From googling a bit is seems that there should be a "Project->Add files recursively..." menu?

xaviersellam replied:
Thank you, I think it works!!
We continue our tasks in order to give you something soon.

xaviersellam commented:
We are trying to change our code: 
- We put the constantparameters as an argument (but difficult to change this) so we think that if we put it in private attribute it will work better. 
- We also change the mcvanillaengine: we finally put the if(constantparameters...) in order to not change the initial class mcvanillaengine
- We are also trying to make a good main
Finally, in ConstantBSprocess, we define parameters like this:
Real S;
Rate r;

But it should be something like this no?:
		Handle<Quote> S;
		Handle<YieldTermStructure> r;
In order to have the same implementation as in the other classes.

We will try this stuffs but if you think that it is not relevant, please tell us. Otherwise, we will send you what we did tomorrow or Thursday.

Thank you,
Nazim and Xavier

lballabio commented:
Real is ok, you don't need Handle.  constantparameter as an attribute is also correct.

xaviersellam commented:
Hi,

We merged what we did in mcvanillaengine, stochasticprocess, mceuropeanengine in european engine only to reduce the number of files that we had. (So don't look at mcvanillaengine and stochasticprocess), I didn't erase them and I can't erase a commit.

We did what we say yesturday: 
we put the constantparameter in private with a set in order to be able to use it.
we added the maturity date on BSprocess
we computed a main.

Please tell us if you see something wrong,
Xavier and Nazim

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,38 @@
+Real newq;
+
+    public:
+ConstantBSProcess(Real S,Rate r,Volatility v,Rate q,Date maturity,
```
I see. I thought you were passing the extracted rates and volatility, as you did before.  If you pass the curves and extract the average values inside the constructor (which is also ok), then you need to pass handles, otherwise the types won't match.

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,38 @@
+class ConstantBSProcess : public StochasticProcess1D {
+    private:
+Real S;
+Rate r;
```
If you extract and store `newr` etc., you don't need to store `r`, `v` and `q`.

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,41 @@
+                maturity=maturity;
+                newr = r->zeroRate(maturity,r-> dayCounter(), Continuous,NoFrequency,true);
+                newq = q->zeroRate(maturity,q-> dayCounter(), Continuous,NoFrequency,true);
+                newv = v->blackVol(maturity, S->value());
```
Here you should use the strike of the option, not the current value of the underlying.

You can pass it from the engine, as you do for the maturity, but it's a bit more complex.  Inside the engine you need to write

    ext::dynamic_pointer_cast<StrikedTypePayoff>(this->arguments_.payoff)->strike()

lballabio commented on project1/blackscholesprocess.cpp:
```diff
@@ -0,0 +1,273 @@
+
```
If you don't use these files, please remove them.  It's a lot less confusing for all of us.

lballabio commented on project1/mceuropeanengine.hpp:
```diff
@@ -61,9 +68,46 @@ namespace QuantLib {
+        private:
+            bool constantparameters = true;
+        public:
+            void setconstantparameters() const {
```
I would have preferred passing it to the constructor, but this is not wrong.  You can leave it if you're more comfortable.

lballabio commented on project1/mcvanillaengine.hpp:
```diff
@@ -0,0 +1,189 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
```
If you don't use this, please remove it to avoid confusion.

xaviersellam commented:
Is this something like this?

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,42 @@
+                boost::shared_ptr<discretization>& d )
+                : StochasticProcess1D(d),S(S), r(r), v(v), q(q){
+                maturity=maturity;
+                strike= ext::dynamic_pointer_cast<StrikedTypePayoff>(this->arguments_.payoff)->strike()
```
You're passing the strike to the constructor, no need to recalculate it.

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,42 @@
+                strike= ext::dynamic_pointer_cast<StrikedTypePayoff>(this->arguments_.payoff)->strike()
+                newr = r->zeroRate(maturity,r-> dayCounter(), Continuous,NoFrequency,true);
+                newq = q->zeroRate(maturity,q-> dayCounter(), Continuous,NoFrequency,true);
+                newv = v->blackVol(maturity, S->value(),true);
```
This would be 
```suggestion
                newv = v->blackVol(maturity, strike, true);
```

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,42 @@
+ConstantBSProcess::ConstantBSProcess(const Handle<Quote> S,const Handle<YieldTermStructure>& r,const Handle<BlackVolTermStructure>& v,const Handle<YieldTermStructure>& q, Date maturity,Real strike,
+                boost::shared_ptr<discretization>& d )
+                : StochasticProcess1D(d),S(S), r(r), v(v), q(q){
+                maturity=maturity;
```
This is assigning the data member to itself.  Either you add `maturity(maturity)` above with `S(S)` etc, or you need to use two different names.  But actually, you don't need to save `maturity`.

lballabio commented on project1/ConstantBSProcess.h:
```diff
@@ -0,0 +1,39 @@
+class ConstantBSProcess : public StochasticProcess1D {
+    private:
+Handle<Quote> S;
+Handle<YieldTermStructure> r;
+Handle<BlackVolTermStructure> v;
+Handle<YieldTermStructure> q;
+Date maturity;
```
You don't need to save r, v, q, maturity and strike.  You won't use them outside the constructor.

lballabio commented on project1/mceuropeanengine.hpp:
```diff
@@ -61,9 +68,46 @@ namespace QuantLib {
+				RNG::make_sequence_generator(dimensions*(grid.size() - 1), seed_);
+
+			if (this->constantparameters) {
+			boost::shared_ptr<GeneralizedBlackScholesProcess> process =
```
Here is where you need to add 

    Real strike= ext::dynamic_pointer_cast<StrikedTypePayoff>(this->arguments_.payoff)->strike();

so you can pass it to the constructor of `ConstantBSProcess`.

xaviersellam replied:
I didn't write exactly what you wrote. Maybe I didn't understand where to put this but there is no error with what I did

lballabio replied:
Ok, what you did works.

xaviersellam commented:
we have some figures and things to explain, do you think it's ok and we can start the report or there are things that remain unconvincing?

lballabio commented on project1/ConstantBSProcess.cpp:
```diff
@@ -0,0 +1,40 @@
+
+ConstantBSProcess::ConstantBSProcess(const Handle<Quote> S,const Handle<YieldTermStructure>& r,const Handle<BlackVolTermStructure>& v,const Handle<YieldTermStructure>& q, Date maturity,Real strike,
+                boost::shared_ptr<discretization>& d )
+                : StochasticProcess1D(d){
```
You're not saving S, that you need in the `x0()` method.
```suggestion
                : StochasticProcess1D(d), S(S) {
```

lballabio commented on project1/main.cpp:
```diff
@@ -1,19 +1,130 @@
+        Date maturity(20, September, 2019);
+        Date settlementDate(20, September, 2018);
+        DayCounter dayCounter = Actual365Fixed();
+
```
By default, the program runs as of today's date, so the option is expired.  If you need to run as of the settlement date, add

    Settings::instance().evaluationDate() = settlementDate;

xaviersellam commented:
The program runs and compiles, we made a few changes!
The only weird thing is that there should a loss of accuracy but the error is approximatively the same for both processes

lballabio commented:
In the case of your main, you're using flat curves so the process is actually already constant even if you're using the general process. You might try using something like ZeroCurve or BlackVarianceCurve instead of FlatForward and BlackConstantVol, so the parameters change in time. Even in that case, though, a European option is simple enough that using the average rates and volatility will probably give similar results.

You would see more difference if you priced a path-dependent option, such as an Asian one. But given the conditions, I wouldn't add more work to this project.
